rpc/gnmi/gnmi-specification.md

```diff
@@ -1484,18 +1484,21 @@ that fewer messages are sent to the client. The advantage of such bundling is
 clearly to reduce the number of bytes on the wire (caused by message overhead);
 however, since only `Notification` messages contain the timestamp at which an
 event occurred, or a sample was taken, such bundling assigns a single timestamp
-for all bundled `Update` values. As such, it has the downside of negatively
-affecting the sample accuracy and freshness to the client, and as a result, on
-the client's ability to react to events on the target.
-
-Since it is not possible for the target to infer whether its clients are
-sensitive to the latency introduced by bundling, if a target implements
-optimizations such that multiple `Update` messages are bundled together,
-it MUST provide an ability to disable this functionality within the
-configuration of the gNMI service. Additionally, a target SHOULD provide means
-by which the operator can control the maximum number of updates that are to be
-bundled into a single message,  This configuration is expected to be implemented
-out-of-band to the gNMI protocol itself.
+for all bundled `Update` values. As such, bundling is primarily useful for
+datasets where a group of leaves are meaningfully conjoined, such as a group
+of leaves atomically applied as a configuration update via a Set call, system
+properties that are effectively static after boot and component inventory data
+including part, model and serial numbers.
+
+For counter and event data where hardware provides precise timestamps, a
+gNMI implementation MUST not obscure access to these timestamps in an
+attempt to provide bundling. In cases where an OpenConfig leaf is derived
+from two or more hardware values with distinct timestamps, an implementation
+SHOULD attempt to provide a consistent and meaningful timestamp that
+introduces minimal error.  This includes attempting to synchronize collection
```

Comment:
This is slightly prescriptive language that perhaps doesn't really tell an implementor what to do. Do we think that having these examples helps with clarity, or will raise more questions?

Reply:
Knowing there are a wide range of hardware implementations and the fact that the OpenConfig interfaces model covers both physical and logical interfaces we will run into situations where a single leaf might be one raw counter in one implementation, but a sum of multiple counters in another.  In the case of multiple counters, they might be retrievable with tight synchronization if located in the same chip, or less so if distributed across multiple chips, or even multiple line cards.  I was attempting to enumerate some of the considerations an implementor should make without being overly prescriptive whilst also conveying the goal of having the computed rates over aggregated counters reflect reality with accuracy. We had one request from a vendor to publish an acceptable error bar, but I worry that arbitrarily picking might preclude an implementation on one or more platforms.  Maybe we ought to have implementors publish an error bar in such situations?

Reply:
```suggestion
introduces minimal error.  This could include approaches such as attempting to synchronize collection
```

How about this clarification of wording that makes it clear that we're not being prescriptive about these approaches but suggesting some considerations.

I agree that having an error bar requirement seems difficult to choose what the value should be -- I mean, why would the consumer want it to be anything other than 0%. Understanding the error bar might be of interest, but I wonder how easy it is to actually quantify?